### PR TITLE
asmjs_sbrk

### PR DESCRIPTION
**Pull request by juj**
Remove variable DYNAMICTOP and replace it with DYNAMICTOP_PTR which points to a memory location on the Emscripten HEAP that contains the ceiling of the dynamically allocated memory area. Remove proxying of sbrk. Relates to https://github.com/kripken/emscripten/issues/4391.

The reason why this is done is two-fold:

- because in future wasm "pure pthreads" mode, pthreads can't jump out of asm, so pthreads would not be able to allocate memory without proxying the requests to the main thread if they happen to need to bump dynamic top
- in the upcoming emscripten_fetch work, the emscripten_fetch worker (which manages XHRs and IndexedDB operations for the whole application) also needs to malloc memory, but having to proxy to main thread to do it would be a big sequential sync point slowdown.

Having the dynamictop variable inside the asm.js heap, which is as a SharedArrayBuffer fully shared between threads and the emscripten_worker, malloc and free can operate properly multithreaded.

I need to run the full tests on this to ensure everything works before merging, but decided to splice this off from the upcoming emscripten_fetch PR altogether to make it easier to review. This PR also implements the `brk` function, which is part of the standard. Not used in this PR itself, but since it's easy to implement, it's nice to support if some compiled code may need it in the future.

### CODE REVIEW

File: src/library.js [hunk 1]
```diff
@@ -428,28 +428,37 @@ LibraryManager.library = {
...
+  // address of the previous top ('break') of the memory area
+  // We control the "dynamic" memory - DYNAMIC_BASE to DYNAMICTOP
+  sbrk__asm: true,
+  sbrk__sig: ['ii'],
+  sbrk: function(increment) {
+    increment = increment|0;
+    var dynamicTop = 0;
 #if USE_PTHREADS
-    if (ENVIRONMENT_IS_PTHREAD) return _emscripten_sync_run_in_main_thread_1({{{ cDefine('EM_PROXIED_SBRK') }}}, bytes);
+    dynamicTop = Atomics_add(HEAPU32, DYNAMICTOP_PTR>>2, increment);
```
Comment (kripken) [inline src/library.js]: the pthreads increment doesn't seem to add padding for aligning, as the other code path does?

File: src/runtime.js [hunk 2]
```diff
@@ -59,15 +59,19 @@ var RuntimeGenerator = {
...
-#if USE_PTHREADS
-    if (typeof ENVIRONMENT_IS_PTHREAD !== 'undefined' && ENVIRONMENT_IS_PTHREAD) throw 'Runtime.dynamicAlloc is not available in pthreads!'; // This is because each worker has its own copy of DYNAMICTOP, of which main thread is authoritative.
-#endif
-    var ret = RuntimeGenerator.alloc(size, 'DYNAMIC');
-    if (SAFE_HEAP) ret += '; if (asm) { Runtime.setDynamicTop(DYNAMICTOP); }';
-    ret += '; if (DYNAMICTOP >= TOTAL_MEMORY) { var success = enlargeMemory(); if (!success) { DYNAMICTOP = ret; ';
-    if (SAFE_HEAP) ret += 'if (asm) { Runtime.setDynamicTop(DYNAMICTOP); }';
-    ret += ' return 0; } }'
+    var ret = '';
+    if (ASSERTIONS) ret += 'assert(DYNAMICTOP_PTR != 0);'; // dynamic area must be ready
```
Comment (kripken) [inline src/runtime.js]: can just assert`(DYNAMICTOP_PTR)`, then it is clear it is also checking it is not `undefined`

File: src/library.js [hunk 3]
```diff
@@ -428,28 +428,37 @@ LibraryManager.library = {
...
-    return ret;  // Previous break location.
+    dynamicTop = (dynamicTop + increment)|0;
+    dynamicTop = ((dynamicTop + 15) & -16)|0;
+#if !USE_PTHREADS
+    HEAPU32[DYNAMICTOP_PTR>>2] = dynamicTop;
+#endif
+    return dynamicTop|0; // TODO: error checking, return -1 on error and set errno.
+  },
+
+  brk__asm: true,
```
Comment (kripken) [inline src/library.js]: what is this new method `brk`? It doesn't seem to be used or tested?

File: src/library.js [hunk 4]
```diff
@@ -428,28 +428,37 @@ LibraryManager.library = {
...
-      var success = self.alloc(bytes);
-      if (!success) return -1 >>> 0; // sbrk failure code
-    }
-    return ret;  // Previous break location.
+    dynamicTop = (dynamicTop + increment)|0;
+    dynamicTop = ((dynamicTop + 15) & -16)|0;
+#if !USE_PTHREADS
+    HEAPU32[DYNAMICTOP_PTR>>2] = dynamicTop;
+#endif
+    return dynamicTop|0; // TODO: error checking, return -1 on error and set errno.
```
Comment (kripken) [inline src/library.js]: i believe this TODO must be done before landing, for memory growth and memory growth failure tests to pass.

**Comment by kripken** [comment]
Nice work, and I like how this removes the previous hacks inside `sbrk` for `dynamicAlloc`.

File: src/library.js (hunk 3 above)
Comment (juj) [inline src/library.js]: `brk` comes from http://man7.org/linux/man-pages/man2/brk.2.html, I just thought to add it in when I found it in the Linux docs, if someone happened to be porting some code that would be using it.

File: src/runtime.js (hunk 2 above)
Comment (juj) [inline src/runtime.js]: yeah, good point

File: src/library.js (hunk 1 above)
Comment (juj) [inline src/library.js]: That's true, thanks. Also looks like this was broken for non-pthreads case, fixed.

File: src/library.js (hunk 4 above)
Comment (juj) [inline src/library.js]: Implemented now.

File: src/library.js [hunk 5]
```diff
@@ -428,28 +428,47 @@ LibraryManager.library = {
...
+  sbrk: function(increment) {
+    increment = increment|0;
+    var dynamicTop = 0;
+    increment = ((increment + 15) & -16)|0;
 #if USE_PTHREADS
-    if (ENVIRONMENT_IS_PTHREAD) return _emscripten_sync_run_in_main_thread_1({{{ cDefine('EM_PROXIED_SBRK') }}}, bytes);
+    dynamicTop = Atomics_add(HEAPU32, DYNAMICTOP_PTR>>2, increment);
+#else
+    dynamicTop = HEAPU32[DYNAMICTOP_PTR>>2]|0;
+    HEAPU32[DYNAMICTOP_PTR>>2] = dynamicTop + increment|0;
```
Comment (kripken) [inline src/library.js]: really just a minor style thing, but `x + y|0` looks odd to me, as the `|0` happens last. putting it so close suggests the opposite. `x + y | 0` (same distance) is what I usually do.

File: src/runtime.js [hunk 6]
```diff
@@ -429,11 +437,11 @@ var Runtime = {
 #endif
 };
 
-Runtime.stackAlloc = unInline('stackAlloc', ['size']);
-Runtime.staticAlloc = unInline('staticAlloc', ['size']);
-Runtime.dynamicAlloc = unInline('dynamicAlloc', ['size']);
-Runtime.alignMemory = unInline('alignMemory', ['size', 'quantum']);
-Runtime.makeBigInt = unInline('makeBigInt', ['low', 'high', 'unsigned']);
+Runtime.stackAlloc = unInline('stackAlloc', ['size'], true);
```
Comment (kripken) [inline src/runtime.js]: why were these changes needed?

File: src/library.js [hunk 7]
```diff
@@ -428,28 +428,47 @@ LibraryManager.library = {
...
+        ___setErrNo({{{ cDefine('ENOMEM') }}});
+        return -1;
+      }
     }
-    return ret;  // Previous break location.
+    return dynamicTop|0;
+  },
+
+  brk__asm: true,
+  brk__sig: ['ii'],
```
Comment (kripken) [inline src/library.js]: if we are adding `brk`, let's add a small test for it? untested code might mean the first time it's silently used by a new codebase it breaks in hard to find ways. or, add a warning on it in `ASSERTIONS` mode (as e.g. we do for some syscalls).

File: src/library.js (hunk 5 above)
Comment (juj) [inline src/library.js]: That's a good point - I actually had to look up the precedence rules once just to verify. I'll add explicit parentheses here so there's no confusion.

Btw, raced with your comments to push a force update on this to fix some test breakages, sorry about that.

**Comment by kripken** [comment]
Please don't add parentheses, though - that clutters it. It's really just one rule to remember in this context, that `x + y | 0` does the `|0` at the end, which is actually what we want, so it's nice to be able to write it without parentheses. And that's how we write it elsewhere.

File: src/runtime.js (hunk 6 above)
Comment (juj) [inline src/runtime.js]: `unInline` was written to always craft the function in the form of `'var ret = ' + string + '; return ret;'`, but the modified `dynamicAlloc` function is not an assignable expression that would fit there, but is  a whole function of its own.

**Comment by juj** [comment]
Oh ok, leaving out the parentheses.

File: src/runtime.js (hunk 6 above)
Comment (kripken) [inline src/runtime.js]: well, you could have avoided touching these lines by making the false value of the new param the old behavior. but either way is fine.

File: src/preamble.js [hunk 8]
```diff
@@ -1004,7 +1006,7 @@ function enlargeMemory() {
 #else
   // TOTAL_MEMORY is the current size of the actual array, and DYNAMICTOP is the new top.
 #if ASSERTIONS
-  assert(DYNAMICTOP >= TOTAL_MEMORY);
+  assert(HEAPU32[DYNAMICTOP_PTR>>2] >= TOTAL_MEMORY);
```
Comment (juj) [inline src/preamble.js]: Btw, here, shouldn't `DYNAMICTOP` need to be strictly larger than `TOTAL_MEMORY`? That is, the address pointed to by dynamic top itself itself is not allocated, so `HEAPU32[HEAPU32[DYNAMICTOP_PTR>>2]]` is not valid indexable memory? If `DYNAMICTOP` == `TOTAL_MEMORY`, that should be a good condition that does not need enlarging?

File: src/preamble.js [hunk 9]
```diff
@@ -1017,9 +1019,9 @@ function enlargeMemory() {
 
   var LIMIT = Math.pow(2, 31); // 2GB is a practical maximum, as we use signed ints as pointers
                                // and JS engines seem unhappy to give us 2GB arrays currently
-  if (DYNAMICTOP >= LIMIT) return false;
+  if (HEAPU32[DYNAMICTOP_PTR>>2] >= LIMIT) return false;
```
Comment (juj) [inline src/preamble.js]: Same here, and in the check below, shouldn't `>=` need to be `>`? (and `<=` be `<` below?)

File: system/lib/split_malloc.cpp [hunk 10]
```diff
@@ -71,7 +71,7 @@ struct Space {
       start = split_memory*index;
     } else {
       // small area in existing chunk 0
-      start = EM_ASM_INT_V({ return (DYNAMICTOP+3)&-4; });
+      start = EM_ASM_INT_V({ return (HEAPU32[DYNAMICTOP_PTR>>2]+3)&-4; });
```
Comment (juj) [inline system/lib/split_malloc.cpp]: This change will need to bump Emscripten version number.

File: tools/ctor_evaller.py [hunk 11]
```diff
@@ -125,7 +125,7 @@ def add_func(asm, func):
 var stackMax = stackTop + totalStack;
 if (stackMax >= totalMemory) throw 'not enough room for stack';
 
-var dynamicTop = stackMax;
+var dynamicTopPtr = stackMax;
```
Comment (juj) [inline tools/ctor_evaller.py]: @kripken: This is actually something I'm not completely certain - what should ctor_evaller do about `DYNAMICTOP_PTR` here?

File: src/preamble.js (hunk 8 above)
Comment (kripken) [inline src/preamble.js]: yes, the assert should be more strict and just `>`.

File: src/preamble.js (hunk 9 above)
Comment (kripken) [inline src/preamble.js]: I suppose this could be `>`, and allow us to allocate exactly 2GB, yes.

Comment (kripken) [inline src/preamble.js]: but below we do want `<=` as we want to keep enlarging to a higher value than is needed, to try to avoid extra growth events

File: system/lib/split_malloc.cpp (hunk 10 above)
Comment (kripken) [inline system/lib/split_malloc.cpp]: hmm, yes. unless you merge it right before the other one that bumps it ;)

File: tools/ctor_evaller.py (hunk 11 above)
Comment (kripken) [inline tools/ctor_evaller.py]: The key thing is that the value of dynamicTop should not change during the ctor evalling. Until now we did that by simply not allowing sbrk to run - it's just not present in the temp .js file we create, so it aborts if called. But with this PR, sbrk is in the asm.js module. So what you want to do is either remove it from the asm.js code, or put an `abort()` on its first line or something like that (the temp .js file we create and eval here isn't valid asm.js, so no need to worry about that).

Comment (kripken) [inline tools/ctor_evaller.py]: and a simple string replace can do that. this is before minification, so just find `function _sbrk(` and e.g. change that to `function _sbrk_should_abort(`.

also remove sbrk from the other places in `ctor_evaller.py` that assumed it was imported

File: src/library.js [hunk 12]
```diff
@@ -428,28 +428,52 @@ LibraryManager.library = {
...
-      self.called = true;
-      assert(Runtime.dynamicAlloc);
-      self.alloc = Runtime.dynamicAlloc;
-      Runtime.dynamicAlloc = function() { abort('cannot dynamically allocate, sbrk now has control') };
-    }
-    var ret = DYNAMICTOP;
-    if (bytes != 0) {
-      var success = self.alloc(bytes);
-      if (!success) return -1 >>> 0; // sbrk failure code
+    oldDynamicTop = Atomics_add(HEAPU32, DYNAMICTOP_PTR>>2, increment);
```
Comment (juj) [inline src/library.js]: Marking down note to self: This needs support for `-s USE_PTHREADS=1 -s ABORTING_MALLOC=0` combination. That unfortunately forces this code to turn into a CAS loop, although that's probably not an issue, since `sbrk` shouldn't be in the critical performance path at all.

File: src/library.js [hunk 13]
```diff
@@ -428,28 +428,52 @@ LibraryManager.library = {
...
+      }
     }
-    return ret;  // Previous break location.
+#endif
+    return oldDynamicTop|0;
+  },
+
+  brk__asm: true,
+  brk__sig: ['ii'],
+  brk: function(addr) {
```
Comment (juj) [inline src/library.js]: This needs the test.

File: src/preamble.js (hunk 9 above)
Comment (juj) [inline src/preamble.js]: Updated these to be stricter. I now realize that 2GB exact is off limits, but 2BG - 16MB should be attainable.  There is some room for improvement in this heuristic so that we could realistically get that as a max cap, although that is not quite yet urgent.

File: src/preamble.js [hunk 14]
```diff
@@ -1470,6 +1480,10 @@ function setF64(ptr, value) {
 
 #endif // USE_PTHREADS
 
+function getTotalMemory() {
```
Comment (juj) [inline src/preamble.js]: This was added so that we can query from inside asm.js `sbrk` to outside world what the current `TOTAL_MEMORY` is, if it got resized. I pondered on making the asm.js module own the `TOTAL_MEMORY` variable as well, but that is not necessary, at least not yet. (Not sure if we get around to doing memory resizing with pthreads support at some point)

File: src/runtime.js [hunk 15]
```diff
@@ -59,15 +59,19 @@ var RuntimeGenerator = {
...
-#if USE_PTHREADS
-    if (typeof ENVIRONMENT_IS_PTHREAD !== 'undefined' && ENVIRONMENT_IS_PTHREAD) throw 'Runtime.dynamicAlloc is not available in pthreads!'; // This is because each worker has its own copy of DYNAMICTOP, of which main thread is authoritative.
-#endif
-    var ret = RuntimeGenerator.alloc(size, 'DYNAMIC');
-    if (SAFE_HEAP) ret += '; if (asm) { Runtime.setDynamicTop(DYNAMICTOP); }';
-    ret += '; if (DYNAMICTOP >= TOTAL_MEMORY) { var success = enlargeMemory(); if (!success) { DYNAMICTOP = ret; ';
-    if (SAFE_HEAP) ret += 'if (asm) { Runtime.setDynamicTop(DYNAMICTOP); }';
-    ret += ' return 0; } }'
+    var ret = '';
+    if (ASSERTIONS) ret += 'assert(DYNAMICTOP_PTR);'; // dynamic area must be ready
```
Comment (juj) [inline src/runtime.js]: The usage of this function turned out to be a bit more subtle than what I thought. `Runtime.dynamicAlloc` is/can be called in two occassions that I understand:

   - when the static memory area has already been sealed (`staticSealed = true`), but the C runtime has not yet been initialized (`runtimeInitialized=false`), i.e. `_malloc` does not yet own allocating memory within the dynamic memory area.
   - when the static memory area has already been sealed (`staticSealed = true`), and the C runtime has been initialized (`runtimeInitialized=true`), but C `_malloc` was not compiled to the program at all, in which case this function owns dynamic memory allocation throughout the program execution. This is very rare, and looks like only `test_cases` hit this case.

I would have liked to add an assert here that would say `assert(!runtimeInitialized or !(C _malloc has control));` but not quite sure if that is possible to check even. In any case, things would critically break if this function gets called when the compiled `_malloc` is already also serving out memory allocations to the dynamic memory area, but not quite sure how to assert that it doesn't happen.

**Comment by juj** [comment]
@kripken: I am now at a point where all tests in `default` and `other` suites pass, except `asm2g.test_eval_ctors`, which I am not sure what is going on there. Perhaps my search-replace increases the size a little and ctor evaller fails to eval any ctors, leaving the final result a net negative of a few characters(?). Would you have a moment to take a peek at that test, and the eval_ctor changes I did in this pull request if those make sense?

Also, before merging, I think I'll need to add a new test for `-s USE_PTHREADS=1 -s ABORTING_MALLOC=0` and make sure that that combo builds and runs ok, since that is a very important build mode to support.

File: system/lib/split_malloc.cpp (hunk 10 above)
Comment (juj) [inline system/lib/split_malloc.cpp]: Ended up bumping the version for both - that'll make regression testing a bit easier in case that will be necessary for someone (let's hope not).

**Comment by kripken** [comment]
What appears to be going on is that the safe-heap changes are triggering a segfault on
````
 if ((dest + bytes | 0) > (HEAPU32[DYNAMICTOP_PTR >> 2] | 0)) segfault();
````
In `asm2g` we test with safe-heap, which is why this doesn't happen anywhere else. And when the segfault happens, it gives up on removing the ctor, so there is no code size improvement.

And the reason for the segfault is that DYNAMICTOP_PTR is given a value of 0, so it looks at address 0 and loads a 0 from there. So e.g. replace
````
var dynamicTopPtr = 0;
````
with
````
var dynamicTopPtr = stackMax;
(new Int32Array(buffer))[dynamicTopPtr >> 2] = stackMax;
````

**Comment by kripken** [comment]
Speaking of which, please add an assert in the ctor evaller js code, after the code runs, that the value in memory there remains `stackMax` or whatever we assign there (as we should not eval ctors that modify dynamicTop).

File: src/library.js (hunk 3 above)
Comment (juj) [inline src/library.js]: Added full implementation and test for brk now.

File: src/preamble.js (hunk 8 above)
Comment (juj) [inline src/preamble.js]: This was updated.

File: tools/ctor_evaller.py (hunk 11 above)
Comment (juj) [inline tools/ctor_evaller.py]: Updated ctor evaller to do this.

File: src/library.js (hunk 12 above)
Comment (juj) [inline src/library.js]: There is now two added tests that stress these kind of cases.

File: src/library.js (hunk 13 above)
Comment (juj) [inline src/library.js]: Added.

**Comment by juj** [comment]
Thanks for the help on ctor evaller, fixed that now.

File: src/library.js [hunk 16]
```diff
@@ -428,28 +428,111 @@ LibraryManager.library = {
...
+    totalMemory = getTotalMemory()|0;
+
+    // Perform a compare-and-swap loop to update the new dynamic top value. This is because
+    // this function can becalled simultaneously in multiple threads.
+    do {
+      oldDynamicTop = Atomics_load(HEAP32, DYNAMICTOP_PTR>>2);
+      newDynamicTop = oldDynamicTop + increment | 0;
+      // Asking to increase dynamic top to a too high value? In pthreads builds we cannot
+      // enlarge memory, so this needs to fail.
+      if (((increment|0) > 0 & (newDynamicTop|0) < (oldDynamicTop|0)) // Detect and fail if we would wrap around signed 32-bit int.
```
Comment (juj) [inline src/library.js]: Since we are dealing with 32-bit signed integers, it is possible to wrap around the allocation, so added explicit checks here.

File: src/library_pthread.js [hunk 17]
```diff
@@ -321,6 +321,10 @@ var LibraryPThread = {
             url: currentScriptUrl,
             buffer: HEAPU8.buffer,
             tempDoublePtr: tempDoublePtr,
+            TOTAL_MEMORY: TOTAL_MEMORY,
+            STATICTOP: STATICTOP,
+            DYNAMIC_BASE: DYNAMIC_BASE,
+            DYNAMICTOP_PTR: DYNAMICTOP_PTR,
```
Comment (juj) [inline src/library_pthread.js]: Here I did a slightly unrelated change. Before, the runtime in pthreads would compute these variables itself, and since it's executing the same .js files and these are static, pthreads would come to the same "conclusion" for these values even when they recompute these four fields. It is conceptually nicer to have pthreads never compute any of these even though they are static, but always have the main runtime pass the values over when constructing the threads workers. so that they will explicitly stay in sync (if something might change in the future).

File: src/pthread-main.js [hunk 18]
```diff
@@ -26,31 +29,32 @@ var Module = {};
...
 }
 Module['print'] = threadPrint;
 Module['printErr'] = threadPrintErr;
-
-// Work around https://bugzilla.mozilla.org/show_bug.cgi?id=1049091
-console = {
-  log: threadPrint,
-  error: threadPrintErr
-};
-
```
Comment (juj) [inline src/pthread-main.js]: The above change removes an old workaround that is no longer needed. Unrelated, but was helpful to do this to ease debugging (removing this causes console logs appear earlier since they don't need to be queued via postMessage)

File: src/pthread-main.js [hunk 19]
```diff
@@ -26,31 +29,32 @@ var Module = {};
...
+    // Initialize the thread-local field(s):
     tempDoublePtr = e.data.tempDoublePtr;
+
+    // Initialize the global "process"-wide fields:
+    buffer = e.data.buffer;
+    TOTAL_MEMORY = e.data.TOTAL_MEMORY;
+    STATICTOP = e.data.STATICTOP;
+    DYNAMIC_BASE = e.data.DYNAMIC_BASE;
+    DYNAMICTOP_PTR = e.data.DYNAMICTOP_PTR;
+
```
Comment (juj) [inline src/pthread-main.js]: Here we now explicitly receive the global system fields from the main thread, instead of recomputing them.

**Comment by juj** [comment]
Phew, everything should now be done. Ran default (all modes), other and browser suite to verify and they all look good. This should now be good for final review and merge.

File: src/preamble.js [hunk 20]
```diff
@@ -933,9 +943,9 @@ function updateGlobalBufferViews() {
   Module['HEAPF64'] = HEAPF64 = new Float64Array(buffer);
 }
 
-var STATIC_BASE = 0, STATICTOP = 0, staticSealed = false; // static area
```
Comment (kripken) [inline src/preamble.js]: i asked this already but either it got lost or i missed the answer. why remove the initializations for all these?

**Comment by kripken** [comment]
lgtm except for my last question.

File: src/preamble.js (hunk 20 above)
Comment (juj) [inline src/preamble.js]: Sorry, missed that one, a good question. The pattern here is that the initialization to ` = 0;` would kill the value that these are initialized to earlier, i.e. in `pthread-main.js` there's

    var STATIC_BASE = 0;
    ...
    STATIC_BASE = <value received from main thread>;
    importScripts('a.js');

so if the file `a.js` has

    var STATIC_BASE = 0;

it would unconditionally override the value that `pthread-main.js` has already intialized it to. `var STATIC_BASE;` doesn't override, but just re-declares and keeps the value if it exists, or if it didn't, it defines the variable and sets it to `undefined`. (at least that's my understanding of the semantics)

I thought of doing the initialization explicitly on the main thread like this:

    var STATIC_BASE, STATICTOP, staticSealed; // static area
    var STACK_BASE, STACKTOP, STACK_MAX; // stack area
    var DYNAMIC_BASE, DYNAMICTOP_PTR; // dynamic area handled by sbrk
    if (!ENVIRONMENT_IS_PTHREAD) {
        STATIC_BASE = STATICTOP = STACK_BASE = STACKTOP = STACK_MAX = DYNAMIC_BASE = DYNAMICTOP_PTR = 0;
        staticSealed = false;
    }

but when running the code without these initialization assignments, I realized everything passes even without these, so figured to just leave it in the kind of "why pay code lines for statements that aren't needed?" fashion. If you think the init to zero is worth having in, certainly can add those. What do you think?

Comment (kripken) [inline src/preamble.js]: I see, thanks.

I think we do still want these inits, as in your code fragment. It matters for the JS engine: if it sees the whole life of the variable can only contain int values, it's better than if it can be undefined as well. When undefined is in the mix its harder to optimize.

Also please add a comment explaining that it would otherwise be initted by the pthreads code.

Comment (juj) [inline src/preamble.js]: Ok, makes sense. Updated with the explicit initialization.

**Comment by juj** [comment]
Thanks for the review! This will build the cornerstone of https://github.com/juj/emscripten/tree/emscripten_fetch, looking towards cleaning that up for a PR next, but before that I'm going to try it out on some test content so that I can get confidence that the api is stable enough to pull in.

**Comment by kripken** [comment]
Great!
